internetee/auction_center
# connect lhv connect

close #https://github.com/internetee/eis_billing_system/issues/93
linked to #https://github.com/internetee/eis_billing_system/pull/94

**What have I done?**
Implemented the functionality for receiving transactions from LHV Connect. Also added support for reference numbers for each user. Now when a user is created, they are assigned a new reference number, as well as a rake task that assigns these new values to all existing users. Also created two tables that store information about transactions - BankStatement and BankTransaction.

Now when information about a completed transaction is received, the billing system checks which service the reference number belongs to. If it turns out that this reference number belongs to a user from the auction, the transaction is forwarded to the auction. In the auction, it searches for invoices of this user that are, first, unpaid, and second, **STRICTLY MATCH THE AMOUNT**. If the amount does not match, then all users with the role of administrator receive an email notification, as well as the content of the transaction itself.

**What needs to be known before starting to test/deploy in production?**

- Firstly, all necessary migrations must be made.
- Secondly, data migration needs to be carried out. This can be done with the command `rake db:migrate:with_data`, if it's necessary to roll back, then the migration rollback is done as follows: `VERSION=20231102091939 rake db:migrate:down:with_data`
- Thirdly, all existing users need to be assigned a reference number through a rake task: `rake reference_no:assign_to_user`

**How to start testing?**

It is necessary to consider the problem comprehensively. This means that first and foremost, you need to ensure that:

- [ ] if the initiator is a registrar, then it is necessary to make sure that transactions towards the registrar still work as before, both in relation to cancelled invoices, unpaid invoices, and paid invoices
- [ ] if it's an auction, then as mentioned in the description, for an unpaid invoice to be marked as paid, the exact amount that belongs to the payment must match the amount that is supposed to come in the transaction
- [ ] You also need to check different cases, such as if there is an unpaid invoice, but the amount received is either less or more.
- [ ] Also, test cases with already paid invoices.
- [ ] Verify that if another user has an invoice for the same amount.
- [ ] Also, check with canceled invoices.
- [ ] And make sure that the email notification is received when it is not possible to match the transaction or invoice that needs to be paid.

**NB! If you need information on how to upload data to LHV Connect to fully test it, please send me a direct message!**
